airbyte-cdk/bulk/core/load/src/integrationTest/kotlin/io/airbyte/cdk/load/mock_integration_test/MockDestinationBackend.kt

```diff
@@ -17,6 +21,62 @@ object MockDestinationBackend {
         getFile(filename).addAll(records)
     }
 
+    fun upsert(
+        filename: String,
+        primaryKey: List<List<String>>,
+        cursor: List<String>,
+        vararg records: OutputRecord
+    ) {
+        fun getField(path: List<String>, record: OutputRecord): AirbyteValue? {
```

Comment:
most (all?) destinations don't actually support traversing a jsonpath to fetch a cursor/PK, but it's easy enough to implement in java :shrug: